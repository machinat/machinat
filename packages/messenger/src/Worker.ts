import crypto from 'crypto';
import fetch from 'node-fetch';
import FormData from 'form-data';
import type { MachinatWorker } from '@machinat/core/engine';
import MachinatQueue, { JobResponse } from '@machinat/core/queue';
import GraphAPIError from './Error';
import type { MessengerJob, MessengerResult, BatchApiRequest } from './types';

type MessengerSendingResponse = JobResponse<MessengerJob, MessengerResult>;

type MessengerQueue = MachinatQueue<MessengerJob, MessengerResult>;

const POST = 'POST';

const REQEST_JSON_HEADERS = { 'Content-Type': 'application/json' };

const appendField = (body: string, key: string, value: string) =>
  `${body.length === 0 ? body : `${body}&`}${key}=${encodeURIComponent(value)}`;

const encodeUriBody = (fields: { [key: string]: any }): string => {
  let body = '';

  for (const key of Object.keys(fields)) {
    const fieldValue = fields[key];

    if (fieldValue !== undefined) {
      body = appendField(
        body,
        key,
        typeof fieldValue === 'string' ? fieldValue : JSON.stringify(fieldValue)
      );
    }
  }

  return body;
};

const formatRequest = (request: BatchApiRequest) =>
  request.method === 'DELETE'
    ? {
        ...request,
        // NOTE: workaround because batch api do not support DELETE with body
        relative_url: request.body
          ? `${request.relative_url}?${encodeUriBody(request.body)}`
          : request.relative_url,
        body: undefined,
      }
    : {
        ...request,
        body: request.body && encodeUriBody(request.body),
      };

const makeRequestName = (channelId: string, count: number) =>
  `${channelId}-${count}`;

const makeFileName = (num: number) => `file_${num}`;

const appendFieldsToForm = (
  form: FormData,
  body: { [key: string]: string | null | undefined }
) => {
  const fields = Object.keys(body);

  for (let k = 0; k < fields.length; k += 1) {
    const field = fields[k];
    const value = body[field];

    if (value) form.append(field, value);
  }

  return form;
};

type TimeoutID = ReturnType<typeof setTimeout>;

export default class MessengerWorker
  implements MachinatWorker<MessengerJob, MessengerResult>
{
  token: string;
  consumeInterval: number;
  private _graphApiEntry: string;
  private _appSecretProof: string | undefined;

  private _started: boolean;
  // @ts-expect-error: keep for later use
  private _isConsuming: boolean;
  private _consumptionTimeoutId: TimeoutID | null;

  constructor(
    accessToken: string,
    consumeInterval: number,
    graphApiVersion: string,
    appSecret: undefined | string
  ) {
    this.token = accessToken;
    this.consumeInterval = consumeInterval;

    this._graphApiEntry = `https://graph.facebook.com/${graphApiVersion}/`;

    this._appSecretProof = appSecret
      ? crypto.createHmac('sha256', appSecret).update(accessToken).digest('hex')
      : undefined;

    this._started = false;
    this._consumptionTimeoutId = null;
  }

  get started(): boolean {
    return this._started;
  }

  start(queue: MessengerQueue): boolean {
    if (this._started) {
      return false;
    }

    if (this.consumeInterval === 0) {
      queue.onJobs(this._listenJobCallback);
    }

    this._started = true;

    this._consume(queue);
    return true;
  }

  stop(queue: MessengerQueue): boolean {
    if (!this._started) {
      return false;
    }

    if (this._consumptionTimeoutId !== null) {
      clearTimeout(this._consumptionTimeoutId);
      this._consumptionTimeoutId = null;
    }

    this._started = false;
    queue.removeJobsListener(this._listenJobCallback);
    return true;
  }

  private _listenJobCallback = this._listenJob.bind(this);

  private _listenJob(queue: MessengerQueue) {
    if (this._started) {
      this._consume(queue);
    }
  }

  private _consumeCallback = this._consume.bind(this);

  private async _consume(queue: MessengerQueue) {
    this._isConsuming = true;
    this._consumptionTimeoutId = null;

    while (queue.length > 0) {
      try {
        // eslint-disable-next-line no-await-in-loop
        await queue.acquire(50, this._executeJobsCallback);
      } catch (e) {
        // leave the error to request side of queue to handle
      }
    }

    this._isConsuming = false;
    if (this._started && this.consumeInterval !== 0) {
      this._consumptionTimeoutId = setTimeout(
        this._consumeCallback,
        this.consumeInterval,
        queue
      );
    }
  }

  private _executeJobsCallback = this._executeJobs.bind(this);

  private async _executeJobs(jobs: MessengerJob[]) {
    const channelSendingRec = new Map();
    let fileCount = 0;
    let filesForm: undefined | FormData;

    const requests = new Array(jobs.length);

    for (let i = 0; i < jobs.length; i += 1) {
      const { request, channelUid, attachmentFileData, attachmentFileInfo } =
        jobs[i];

      if (channelUid !== undefined) {
        // keep the order of requests per channel
        let count = channelSendingRec.get(channelUid);
        if (count !== undefined) {
          request.depends_on = makeRequestName(channelUid, count);
          count += 1;
        } else {
          count = 1;
        }

        channelSendingRec.set(channelUid, count);
        request.name = makeRequestName(channelUid, count);
      }

      request.omit_response_on_success = false;

      // if binary data attached, use from-data and add the file field
      if (attachmentFileData !== undefined) {
        if (filesForm === undefined) {
          filesForm = new FormData();
        }

        const filename = makeFileName(fileCount);
        fileCount += 1;

        filesForm.append(filename, attachmentFileData, attachmentFileInfo);
        request.attached_files = filename;
      }

      requests[i] = formatRequest(request);
    }

    const batchBody = {
      access_token: this.token,
      appsecret_proof: this._appSecretProof || undefined,
      batch: JSON.stringify(requests),
    };

    // use formdata if files attached, otherwise json
    const body =
      filesForm !== undefined
        ? appendFieldsToForm(filesForm, batchBody)
        : JSON.stringify(batchBody);

    const apiResponse = await fetch(this._graphApiEntry, {
      method: POST,
      headers: filesForm !== undefined ? undefined : REQEST_JSON_HEADERS,
      body,
    });

    // if batch respond with error, throw it
    const apiBody = await apiResponse.json();
    if (!apiResponse.ok) {
      throw new GraphAPIError(apiBody);
    }

    const jobResponses: MessengerSendingResponse[] = new Array(apiBody.length);

    for (let i = 0; i < jobResponses.length; i += 1) {
      const result = apiBody[i];
      result.body = JSON.parse(result.body);

      const success = result.code >= 200 && result.code < 300;

      jobResponses[i] = success // NOTE: to help flow recognize which case it is
        ? {
            success: true,
            result,
            error: undefined,
            job: jobs[i],
          }
        : {
            success: false,
            result,
            error: new GraphAPIError(result.body),
            job: jobs[i],
          };
    }

    return jobResponses;
  }
}
