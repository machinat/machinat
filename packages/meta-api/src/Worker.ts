import crypto from 'crypto';
import fetch from 'node-fetch';
import FormData from 'form-data';
import type { ChannelSettingsAccessor, SociablyChannel } from '@sociably/core';
import type { SociablyWorker } from '@sociably/core/engine';
import SociablyQueue, { JobResponse } from '@sociably/core/queue';
import GraphAPIError from './Error';
import formatBatchRequest from './utils/formatBatchRequest';
import appendFormFields from './utils/appendFormFields';
import getObjectValueByPath from './utils/getObjectValueByPath';
import type { MetaApiJob, MetaApiResult, MetaBatchRequest } from './types';

type MetaSendingSettings = {
  accessToken: string;
};
type MetaSettingsAccessor = ChannelSettingsAccessor<
  SociablyChannel,
  MetaSendingSettings
>;

type MetaApiQueue = SociablyQueue<MetaApiJob, MetaApiResult>;
type TimeoutID = ReturnType<typeof setTimeout>;

const POST = 'POST';
const REQEST_JSON_HEADERS = { 'Content-Type': 'application/json' };

const makeRequestName = (key: string, count: number) => `${key}-${count}`;

const makeFileName = (num: number) => `file_${num}`;

class MetaApiWorker implements SociablyWorker<MetaApiJob, MetaApiResult> {
  consumeInterval: number;
  private _settingsAccessor: MetaSettingsAccessor;
  private _appSecret: undefined | string;
  private _graphApiEntry: string;

  private _started: boolean;
  // @ts-expect-error: keep for later use
  private _isConsuming: boolean;
  private _consumptionTimeoutId: TimeoutID | null;

  private _resultCache: Map<string, MetaApiResult>;

  constructor(
    settingsAccessor: MetaSettingsAccessor,
    appSecret: undefined | string,
    graphApiVersion: string,
    consumeInterval: number
  ) {
    this._settingsAccessor = settingsAccessor;
    this.consumeInterval = consumeInterval;

    this._appSecret = appSecret;
    this._graphApiEntry = `https://graph.facebook.com/${graphApiVersion}/`;

    this._started = false;
    this._consumptionTimeoutId = null;
    this._resultCache = new Map();
  }

  get started(): boolean {
    return this._started;
  }

  start(queue: MetaApiQueue): boolean {
    if (this._started) {
      return false;
    }

    if (this.consumeInterval === 0) {
      queue.onJobs(this._listenJobCallback);
    }

    this._started = true;

    this._consume(queue);
    return true;
  }

  stop(queue: MetaApiQueue): boolean {
    if (!this._started) {
      return false;
    }

    if (this._consumptionTimeoutId !== null) {
      clearTimeout(this._consumptionTimeoutId);
      this._consumptionTimeoutId = null;
    }

    this._started = false;
    queue.removeJobsListener(this._listenJobCallback);
    return true;
  }

  private _createAppSecretProof(accessToken: string) {
    const appSecret = this._appSecret;
    return appSecret
      ? crypto.createHmac('sha256', appSecret).update(accessToken).digest('hex')
      : undefined;
  }

  private _listenJobCallback = this._listenJob.bind(this);
  private _listenJob(queue: MetaApiQueue) {
    if (this._started) {
      this._consume(queue);
    }
  }

  private _consumeCallback = this._consume.bind(this);
  private async _consume(queue: MetaApiQueue) {
    this._isConsuming = true;
    this._consumptionTimeoutId = null;

    while (queue.length > 0) {
      try {
        // eslint-disable-next-line no-await-in-loop
        await queue.acquire(50, this._executeJobsCallback);
      } catch {
        // leave the error to request side of queue
      }
    }

    this._isConsuming = false;
    if (this._started && this.consumeInterval !== 0) {
      this._consumptionTimeoutId = setTimeout(
        this._consumeCallback,
        this.consumeInterval,
        queue
      );
    }
  }

  private async _getAccessTokensOfChannels(
    channels: SociablyChannel[]
  ): Promise<Map<string, string>> {
    const uniqChannelsMap = new Map(channels.map((chan) => [chan.uid, chan]));
    const uniqChannels = [...uniqChannelsMap.values()];

    const settings = await this._settingsAccessor.getChannelSettingsBatch(
      uniqChannels
    );
    return new Map(
      uniqChannels
        .map((chan, i) => [chan.uid, settings[i]?.accessToken])
        .filter((tokenPair): tokenPair is [string, string] => !!tokenPair[1])
    );
  }

  private _executeJobsCallback = this._executeJobs.bind(this);
  private async _executeJobs(jobs: MetaApiJob[]) {
    const sendingKeysCounts = new Map<string, number>();
    const resultRegistry = new Map<string, { name: string; idx: number }>();
    let fileCount = 0;
    let filesForm: undefined | FormData;

    const jobResponses: JobResponse<MetaApiJob, MetaApiResult>[] = new Array(
      jobs.length
    );
    const batchRequests: MetaBatchRequest[] = [];

    const channelTokenMap = await this._getAccessTokensOfChannels(
      jobs.map((job) => job.channel)
    );

    for (let i = 0; i < jobs.length; i += 1) {
      const job = jobs[i];
      const {
        request: requestInput,
        key,
        file,
        registerResult,
        consumeResult,
        channel,
      } = job;

      const accessToken = channelTokenMap.get(channel.uid);
      if (!accessToken) {
        jobResponses[i] = {
          success: false,
          job,
          result: { code: 0, headers: {}, body: {} },
          error: new Error(
            `No access token available for channel ${channel.uid}`
          ),
        };
      } else {
        let request = requestInput;

        // if it requires previous results to accomplish the request
        if (consumeResult) {
          const { keys: consummingKeys, accomplishRequest } = consumeResult;

          const getPreviousResult = (requestKey, path) => {
            const currentBatchRegistered = resultRegistry.get(requestKey);
            if (currentBatchRegistered) {
              return `{result=${currentBatchRegistered.name}:${path}}`;
            }
            const lastBatchRegistered = this._resultCache.get(requestKey);
            if (lastBatchRegistered) {
              return getObjectValueByPath(lastBatchRegistered.body, path);
            }
            return null;
          };
          request = accomplishRequest(
            requestInput,
            consummingKeys,
            getPreviousResult
          );
        }

        const batchRequest = formatBatchRequest(request, accessToken);

        // to keep the order of requests with the same key
        if (key) {
          let count = sendingKeysCounts.get(key);
          if (count !== undefined) {
            batchRequest.depends_on = makeRequestName(key, count);
            count += 1;
          } else {
            count = 1;
          }

          sendingKeysCounts.set(key, count);
          batchRequest.name = makeRequestName(key, count);
        }

        // register to use/cache the result
        if (registerResult) {
          if (!batchRequest.name) {
            batchRequest.name = makeRequestName('#request', i);
          }
          resultRegistry.set(registerResult, {
            name: batchRequest.name,
            idx: i,
          });
        }

        // if binary data attached, use from-data and add the file field
        if (file) {
          if (filesForm === undefined) {
            filesForm = new FormData();
          }

          const filename = makeFileName(fileCount);
          fileCount += 1;

          filesForm.append(filename, file.data, file.info);
          batchRequest.attached_files = filename;
        }

        batchRequest.omit_response_on_success = false;
        batchRequests.push(batchRequest);
      }
    }

    const fallbackToken = channelTokenMap.values().next().value as string;
    const batchBody = {
      access_token: fallbackToken,
      appsecret_proof: this._createAppSecretProof(fallbackToken),
      batch: JSON.stringify(batchRequests),
    };

    // use formdata if files attached, otherwise JSON
    const body =
      filesForm !== undefined
        ? appendFormFields(filesForm, batchBody)
        : JSON.stringify(batchBody);

    const apiRes = await fetch(this._graphApiEntry, {
      method: POST,
      headers: filesForm !== undefined ? undefined : REQEST_JSON_HEADERS,
      body,
    });

    // unexpected batch API error, throw it
    const apiBody = await apiRes.json();
    if (!apiRes.ok) {
      throw new GraphAPIError(apiBody);
    }

    // cache result if it's registered in this batch but not consumed
    this._resultCache.clear();
    for (const [consummingKey, { idx }] of resultRegistry.entries()) {
      this._resultCache.set(consummingKey, apiBody[idx]);
    }

    // collect batch results
    const results = [...apiBody];
    for (let i = 0; i < jobResponses.length; i += 1) {
      if (!jobResponses[i]) {
        const result = results.shift();
        result.body = JSON.parse(result.body);

        jobResponses[i] =
          result.code >= 200 && result.code < 300
            ? {
                success: true,
                result,
                error: undefined,
                job: jobs[i],
              }
            : {
                success: false,
                result,
                error: new GraphAPIError(result.body),
                job: jobs[i],
              };
      }
    }
    return jobResponses;
  }
}

export default MetaApiWorker;
