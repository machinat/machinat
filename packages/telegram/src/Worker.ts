import fetch, { Response } from 'node-fetch';
import FormData from 'form-data';
import type { SociablyWorker } from '@sociably/core/engine';
import Queue from '@sociably/core/queue';
import TelegramUser from './User';
import { BotSettingsAccessorI } from './interface';
import type { TelegramJob, TelegramResult, UploadingFile } from './types';
import TelegramApiError from './Error';

type TelegramJobQueue = Queue<TelegramJob, TelegramResult>;

const API_HOST = 'https://api.telegram.org';

const makeBotApiEntry = (botToken: string) => `${API_HOST}/bot${botToken}`;

const requestBotApi = async (
  botToken: string,
  method: string,
  parameters: { [k: string]: unknown },
  files: undefined | UploadingFile[]
): Promise<TelegramResult> => {
  const botApiEntry = makeBotApiEntry(botToken);
  let response: Response;

  if (files && files.length > 0) {
    const form = new FormData();
    Object.entries(parameters).forEach(([key, value]) => {
      form.append(key, JSON.stringify(value));
    });

    files.forEach(({ fieldName, data, info: fileInfo }) => {
      form.append(fieldName, data, fileInfo);
    });

    response = await fetch(`${botApiEntry}/${method}`, {
      method: 'POST',
      body: form,
      headers: form.getHeaders(),
    });
  } else {
    response = await fetch(`${botApiEntry}/${method}`, {
      method: 'POST',
      body: JSON.stringify(parameters),
      headers: { 'Content-Type': 'application/json' },
    });
  }

  const result = await response.json();

  if (!result.ok) {
    throw new TelegramApiError(result);
  }

  return result;
};

export default class TelegramWorker
  implements SociablyWorker<TelegramJob, TelegramResult>
{
  private _settingsAccessor: BotSettingsAccessorI;
  private _started: boolean;
  private _lockedKeys: Set<string>;

  connectionCount: number;
  maxConnections: number;

  constructor(settingsAccessor: BotSettingsAccessorI, maxConnections: number) {
    this._settingsAccessor = settingsAccessor;
    this.connectionCount = 0;
    this.maxConnections = maxConnections;
    this._lockedKeys = new Set();
    this._started = false;
  }

  get started(): boolean {
    return this._started;
  }

  start(queue: TelegramJobQueue): boolean {
    if (this._started) {
      return false;
    }

    queue.onJobs(this._consumeCallback);
    this._started = true;

    this._consume(queue);
    return true;
  }

  stop(queue: TelegramJobQueue): boolean {
    if (!this._started) {
      return false;
    }

    this._started = false;
    queue.removeJobsListener(this._consumeCallback);
    return true;
  }

  private _consumeCallback = this._consume.bind(this);

  private _consume(queue: Queue<TelegramJob, TelegramResult>) {
    const { _lockedKeys: lockedIds, maxConnections } = this;

    for (let i = 0; i < queue.length; ) {
      if (this.connectionCount >= maxConnections) {
        break;
      }

      const { key } = queue.peekAt(i) as TelegramJob;
      if (key !== undefined && lockedIds.has(key)) {
        i += 1;
      } else {
        this._consumeJobAt(queue, i, key);

        this.connectionCount += 1;

        if (key !== undefined) {
          lockedIds.add(key);
        }
      }
    }
  }

  private async _consumeJobAt(
    queue: Queue<TelegramJob, TelegramResult>,
    idx: number,
    executionKey: undefined | string
  ) {
    try {
      await queue.acquireAt(idx, 1, this._executeJobCallback);
    } catch {
      // NOTE: leave the error to the request side
    } finally {
      this.connectionCount -= 1;

      if (executionKey !== undefined) {
        this._lockedKeys.delete(executionKey);
      }

      if (this._started) {
        this._consume(queue);
      }
    }
  }

  private _executeJobCallback: typeof this._executeJob =
    this._executeJob.bind(this);

  private async _executeJob([job]: TelegramJob[]) {
    const { method, params, files, agentId } = job;
    const botSettings = await this._settingsAccessor.getChannelSettings(
      new TelegramUser(agentId, true)
    );
    if (!botSettings) {
      throw new Error(`Agent bot "${agentId}" not registered`);
    }

    const result = await requestBotApi(
      botSettings.botToken,
      method,
      params,
      files
    );

    return [{ success: true as const, result, job, error: undefined }];
  }
}
