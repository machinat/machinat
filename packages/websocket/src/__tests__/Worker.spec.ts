import { moxy } from '@moxyjs/moxy';
import Queue from '@sociably/core/queue';
import type { WebSocketServer } from '../Server.js';
import Worker from '../Worker.js';
import type { WebSocketJob, WebSocketResult } from '../types.js';

const server = moxy<WebSocketServer<null, null>>({
  dispatch: () => Promise.resolve(null),
  disconnect: () => Promise.resolve(false),
} as never);

const queue = new Queue<WebSocketJob, WebSocketResult>();

beforeEach(() => {
  server.mock.reset();
});

it('subscribe to only one queue', () => {
  const worker = new Worker(server);

  expect(worker.start(queue)).toBe(true);
  expect(worker.start(queue)).toBe(false);
  expect(worker.stop(queue)).toBe(true);
  expect(worker.stop(queue)).toBe(false);
});

it('work', async () => {
  const worker = new Worker(server);
  worker.start(queue);

  const jobs = [
    {
      target: {
        type: 'connection' as const,
        serverId: '#server',
        id: '#conn1',
      },
      values: [{ type: 'foo', payload: '1' }],
    },
    {
      target: { type: 'topic' as const, key: 'somthing' },
      values: [
        { type: 'foo', payload: '2' },
        { type: 'bar', category: 'baz', payload: '3' },
      ],
    },
    {
      target: {
        type: 'connection' as const,
        serverId: '#server',
        id: '#conn2',
      },
      values: [{ type: 'foo', payload: '4' }],
    },
  ];

  const broadcastResult = [
    ['A', 'B', 'C', 'D'],
    ['A', 'B', 'C', 'D'],
    ['A'],
    [],
  ];

  let r = 0;
  server.dispatch.mock.fake(() => broadcastResult[r++]); // eslint-disable-line no-plusplus

  await expect(queue.executeJobs(jobs)).resolves.toMatchSnapshot();

  expect(server.dispatch).toHaveBeenCalledTimes(3);
  for (const [i, { target, values }] of jobs.entries()) {
    expect(server.dispatch).toHaveBeenNthCalledWith(i + 1, {
      target,
      values,
    });
  }
});
