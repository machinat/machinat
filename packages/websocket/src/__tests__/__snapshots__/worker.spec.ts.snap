// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`work 1`] = `
Object {
  "batch": Array [
    Object {
      "error": undefined,
      "job": Object {
        "target": Object {
          "id": "#conn1",
          "serverId": "#server",
          "type": "connection",
        },
        "values": Array [
          Object {
            "payload": "1",
            "type": "foo",
          },
        ],
      },
      "result": Object {
        "connections": Array [
          "A",
          "B",
          "C",
          "D",
        ],
      },
      "success": true,
    },
    Object {
      "error": undefined,
      "job": Object {
        "target": Object {
          "key": "somthing",
          "type": "topic",
        },
        "values": Array [
          Object {
            "payload": "2",
            "type": "foo",
          },
          Object {
            "category": "baz",
            "payload": "3",
            "type": "bar",
          },
        ],
      },
      "result": Object {
        "connections": Array [
          "A",
          "B",
          "C",
          "D",
        ],
      },
      "success": true,
    },
    Object {
      "error": undefined,
      "job": Object {
        "target": Object {
          "id": "#conn2",
          "serverId": "#server",
          "type": "connection",
        },
        "values": Array [
          Object {
            "payload": "4",
            "type": "foo",
          },
        ],
      },
      "result": Object {
        "connections": Array [
          "A",
        ],
      },
      "success": true,
    },
  ],
  "errors": null,
  "success": true,
}
`;
